File: cronosFacialIOS/CronosFacial/ViewModels/MainViewModel.swift
//
//  MainViewModel.swift
//  CronosFacial
//
//  MVVM ViewModel for main screen (equivalent to Android MainViewModel)
//

import Foundation
import Combine
import AVFoundation

/// UI State for the main screen
struct MainUiState {
    var isTracking: Bool = false
    var sessionData: [FaceSessionData] = []
    var engagementStates: [(timestamp: Int64, state: EngagementState)] = []
    var latestEmotion: String = "Neutral"
    var latestEngagement: EngagementState = .unknown
    var message: String? = nil
    var isLoading: Bool = false
    var error: String? = nil
    var hasCameraPermission: Bool = false
}

/// Main ViewModel managing facial analysis session state
/// Equivalent to Android's MainViewModel with StateFlow
class MainViewModel: ObservableObject {
    @Published private(set) var uiState = MainUiState()
    
    private let repository: FacialAnalysisRepository
    private var cancellables = Set<AnyCancellable>()
    
    init(repository: FacialAnalysisRepository) {
        self.repository = repository
        checkCameraPermission()
    }
    
    // MARK: - Camera Permission
    
    func checkCameraPermission() {
        switch AVCaptureDevice.authorizationStatus(for: .video) {
        case .authorized:
            uiState.hasCameraPermission = true
        case .notDetermined:
            requestCameraPermission()
        case .denied, .restricted:
            uiState.hasCameraPermission = false
        @unknown default:
            uiState.hasCameraPermission = false
        }
    }
    
    func requestCameraPermission() {
        AVCaptureDevice.requestAccess(for: .video) { [weak self] granted in
            DispatchQueue.main.async {
                self?.uiState.hasCameraPermission = granted
            }
        }
    }
    
    // MARK: - Tracking Session
    
    /// Start a new tracking session
    func startTracking() {
        uiState.isTracking = true
        uiState.sessionData = []
        uiState.engagementStates = []
        uiState.latestEmotion = "Neutral"
        uiState.latestEngagement = .unknown
        uiState.message = nil
        uiState.error = nil
        
        // Start AI/ML session
        repository.startAiMlSession()
    }
    
    /// Stop the current tracking session and submit data
    func stopTracking() {
        uiState.isTracking = false
        uiState.isLoading = true
        
        Task {
            let result = await repository.submitAllSessionData(
                sessionData: uiState.sessionData,
                engagementStates: uiState.engagementStates
            )
            
            await MainActor.run {
                switch result {
                case .success(let message):
                    uiState.message = message
                    uiState.isLoading = false
                    uiState.error = nil
                    
                case .failure(let error):
                    uiState.error = error.localizedDescription
                    uiState.isLoading = false
                    
                case .loading:
                    break
                }
            }
        }
    }
    
    /// Toggle tracking on/off
    func toggleTracking() {
        if uiState.isTracking {
            stopTracking()
        } else {
            startTracking()
        }
    }
    
    // MARK: - Data Handling
    
    /// Handle new face data from camera
    func onFaceData(_ data: FaceSessionData) {
        guard uiState.isTracking else { return }
        
        uiState.sessionData.append(data)
        uiState.latestEmotion = data.emotion
        
        // Stream to AI/ML backend
        Task {
            let landmarks = [
                FacialLandmark(
                    x: 0.5, y: 0.5, z: 0.0,
                    confidence: 0.8,
                    landmarkType: "face_center"
                )
            ]
            
            _ = await repository.streamFacialData(
                landmarks: landmarks,
                emotion: data.emotion,
                gaze: data.gaze,
                engagement: data.engagement,
                confidence: 0.8
            )
        }
    }
    
    /// Handle new engagement state
    func onEngagement(_ engagement: EngagementState) {
        guard uiState.isTracking else { return }
        
        let timestamp = Int64(Date().timeIntervalSince1970 * 1000)
        uiState.engagementStates.append((timestamp, engagement))
        uiState.latestEngagement = engagement
    }
    
    // MARK: - Message Handling
    
    /// Clear displayed message
    func clearMessage() {
        uiState.message = nil
    }
    
    /// Clear displayed error
    func clearError() {
        uiState.error = nil
    }
}

File: cronosFacial/app/src/main/java/com/cronosedx/cronosfacial/viewmodel/MainViewModel.kt
package com.cronosedx.cronosfacial.viewmodel

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.cronosedx.cronosfacial.common.Result
import com.cronosedx.cronosfacial.model.EngagementState
import com.cronosedx.cronosfacial.network.FacialLandmark
import com.cronosedx.cronosfacial.repository.FacialAnalysisRepository
import com.cronosedx.cronosfacial.ui.FaceSessionData
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import javax.inject.Inject

/**
 * UI state for the main screen
 */
data class MainUiState(
    val isTracking: Boolean = false,
    val sessionData: List<FaceSessionData> = emptyList(),
    val engagementStates: List<Pair<Long, EngagementState>> = emptyList(),
    val latestEmotion: String = "Neutral",
    val latestEngagement: EngagementState = EngagementState.Unknown,
    val message: String? = null,
    val isLoading: Boolean = false,
    val error: String? = null
)

/**
 * ViewModel for the main screen.
 * 
 * Manages the state and business logic for facial analysis sessions,
 * including camera tracking, data collection, and network submissions.
 */
@HiltViewModel
class MainViewModel @Inject constructor(
    private val repository: FacialAnalysisRepository
) : ViewModel() {
    
    private val _uiState = MutableStateFlow(MainUiState())
    val uiState: StateFlow<MainUiState> = _uiState.asStateFlow()
    
    /**
     * Start a new tracking session
     */
    fun startTracking() {
        _uiState.value = _uiState.value.copy(
            isTracking = true,
            sessionData = emptyList(),
            engagementStates = emptyList(),
            latestEmotion = "Neutral",
            latestEngagement = EngagementState.Unknown,
            message = null,
            error = null
        )
        
        // Start AI/ML session
        repository.startAiMlSession()
    }
    
    /**
     * Stop the current tracking session and submit data
     */
    fun stopTracking() {
        _uiState.value = _uiState.value.copy(
            isTracking = false,
            isLoading = true
        )
        
        viewModelScope.launch {
            val result = repository.submitAllSessionData(
                _uiState.value.sessionData,
                _uiState.value.engagementStates
            )
            
            when (result) {
                is Result.Success -> {
                    _uiState.value = _uiState.value.copy(
                        message = result.data,
                        isLoading = false,
                        error = null
                    )
                }
                is Result.Error -> {
                    _uiState.value = _uiState.value.copy(
                        error = result.message ?: result.exception.localizedMessage,
                        isLoading = false
                    )
                }
                is Result.Loading -> {
                    // Already in loading state
                }
            }
        }
    }
    
    /**
     * Toggle tracking on/off
     */
    fun toggleTracking() {
        if (_uiState.value.isTracking) {
            stopTracking()
        } else {
            startTracking()
        }
    }
    
    /**
     * Handle new face data from camera
     */
    fun onFaceData(data: FaceSessionData) {
        if (!_uiState.value.isTracking) return
        
        _uiState.value = _uiState.value.copy(
            sessionData = _uiState.value.sessionData + data,
            latestEmotion = data.emotion
        )
        
        // Stream to AI/ML backend
        viewModelScope.launch {
            val landmarks = listOf(
                FacialLandmark(
                    x = 0.5f, y = 0.5f, z = 0.0f,
                    confidence = 0.8f,
                    landmarkType = "face_center"
                )
            )
            
            repository.streamFacialData(
                landmarks = landmarks,
                emotion = data.emotion,
                gaze = data.gaze,
                engagement = data.engagement,
                confidence = 0.8f
            )
        }
    }
    
    /**
     * Handle new engagement state
     */
    fun onEngagement(engagement: EngagementState) {
        if (!_uiState.value.isTracking) return
        
        val timestamp = System.currentTimeMillis()
        _uiState.value = _uiState.value.copy(
            engagementStates = _uiState.value.engagementStates + (timestamp to engagement),
            latestEngagement = engagement
        )
    }
    
    /**
     * Clear any displayed message
     */
    fun clearMessage() {
        _uiState.value = _uiState.value.copy(message = null)
    }
    
    /**
     * Clear any displayed error
     */
    fun clearError() {
        _uiState.value = _uiState.value.copy(error = null)
    }
}
